README.md

```diff
@@ -1,2 +1,454 @@
+
+While some resources are centralized (e.g. access to a server), others are unbound from specific locations (e.g. a CRDT), in which case it will take longer for the revocation to propagate.
+
+Every resource type SHOULD have a canonical location where its revocations are kept. This list is non-exclusive, and revocation messages MAY be gossiped between peers in a network to share this information more quickly.
+
+It is RECOMMENDED that the canonical revocation store be kept at as close to (or inside) the resource it is about as possible. For example, the WebNative File System maintains a Merkle tree of revoked CIDs at a well-known path. Another example is a centralized server could have an endpoint that lists the revoked UCANs by CID.
+
+Revocations MUST be irreversible. If the revocation was issued in error, a unique UCAN MAY be issued (e.g. by updating the nonce, or changing the time bounds). This prevents confusion as the revocation moves through the network, and makes revocation stores append-only and highly amenable to caching.
+
+A revocation message MUST conform to the following format:
+
+``` js
+{
+  "iss": did,
+  "revoke": ucanCid,
+  "challenge": sign(did.privateKey, `REVOKE:${ucanCid}`)
+}
+```
+
+This format makes it easy to select the relevant UCAN, confirm that the issuer is in the proof chain for some or all capabilities, and validate that the revocation itself was signed by the same issuer. 
+
+Any other witnesses in the selected UCAN not issued by the same DID as the revocation issuer MUST be treated as valid.
+
+Revocations MAY be deleted once the UCAN they reference expires or otherwise becomes invalid via its proactive mechanisms.
+
+## 5.8 Backwards Compatibility
+
+A UCAN validator MAY implement backawrds compatibility with previous version of UCAN. Delegated UCANs MUST be of equal or higher version than their proofs. For example, a v0.8.0 UCAN that includes witnesses that are separately v0.8.0, v0.7.0, and v0.5.0 MAY be considered valid. A v0.5.0 that has a UCAN v0.8.0 witness MUST NOT be considered valid.
+
+# 6. Implementation Recommendations
+
+## 6.1 UCAN Store
+
+A validator MAY keep a local store of UCANs that it has received. UCANs are immutable, but also time bound, so this store MAY evict expired or revoked UCANs.
+
+This store MAY be indexed by CID (content addressing). Multiple indices built on top of this store MAY be used to improve capability search or selection performance.
+
+## 6.2 Memoized Validation
+
+Aside from revocation, UCAN validation is an idempotent action. Marking a CID as valid acts as memoization, obviating the need to check the entire structure on every validation. This extends to distinct UCANs that share a witness: if the witness was previously checked and is not revoked, it is RECOMMENDED to immediately consider it valid.
```

Comment:
I think there is a bit of a disconnect here. On one hand you don't want to or can validate all of the claims in the UCAN chain, meaning that you only verify specific claims, but on the other hand you want to cache validations which you can not really do unless you have validated all the claims.

I'm not sure what I'm exactly asking here, but I feel like this odd to be captured somewhere to avoid UCAN deployments from mistakenly caching partial validations.

Reply:
> but on the other hand you want to cache validations which you can not really do unless you have validated all the claims.

You have lots of options here. For example, you can lazily process the UCAN at invocation time to speed up the request/response, and then do a full validation out of band and cache the results.

Reply:
Good point though, I'll clarify 👍